File: main-server/Services/ChunkAssembly.cs
using System.Collections.Concurrent;
using MainServer.Models;

namespace MainServer.Services;

public sealed class ChunkAssembly
{
    private readonly ConcurrentDictionary<int, string> _chunks = new();
    private int? _finalRequestNumber;

    // 1번 청크에만 실리지만, 실제로 쓰는 시점(마지막 청크)까지 들고 있어야 해서 여기 보관.
    public int? MessageCount { get; private set; }
    public List<HistoryTurn>? History { get; private set; }

    public void Add(
        int requestNumber, string messageChunk, int? finalRequestNumber,
        int? messageCount = null, List<HistoryTurn>? history = null)
    {
        _chunks[requestNumber] = messageChunk;
        if (finalRequestNumber is not null)
        {
            _finalRequestNumber = finalRequestNumber;
        }
        if (messageCount is not null)
        {
            MessageCount = messageCount;
            History = history;
        }
    }

    public bool IsReady(bool isComplete)
    {
        if (!isComplete || _finalRequestNumber is null)
        {
            return false;
        }

        for (var i = 1; i <= _finalRequestNumber; i++)
        {
            if (!_chunks.ContainsKey(i))
            {
                return false;
            }
        }
        return true;
    }

    public string Assemble()
    {
        return string.Concat(Enumerable.Range(1, _finalRequestNumber!.Value).Select(i => _chunks[i]));
    }
}

// session_id별로 진행 중인 청크 조립을 저장해 두는 곳 (MainServer 자신의 DB 저장용 재조립)
public sealed class ChunkAssemblyStore
{
    private readonly ConcurrentDictionary<string, ChunkAssembly> _bySession = new();

    public ChunkAssembly GetOrAdd(string sessionId) => _bySession.GetOrAdd(sessionId, _ => new ChunkAssembly());

    public void Remove(string sessionId) => _bySession.TryRemove(sessionId, out _);
}

File: rag-server/app/chunk_assembly.py
"""청크(조각)로 나뉘어 들어오는 요청을 session_id 기준으로 모아 완성 여부를 판단한다."""


class ChunkAssembly:
    def __init__(self):
        self._chunks: dict[int, str] = {}
        self._final_request_number: int | None = None

    def add(self, request_number: int, message_chunk: str, final_request_number: int | None) -> None:
        self._chunks[request_number] = message_chunk
        if final_request_number is not None:
            self._final_request_number = final_request_number

    def is_ready(self, is_complete: bool) -> bool:
        if not is_complete or self._final_request_number is None:
            return False
        # 1..final_request_number까지 빠진 번호가 없어야 완성으로 본다
        return all(i in self._chunks for i in range(1, self._final_request_number + 1))

    def assemble(self) -> str:
        return "".join(self._chunks[i] for i in range(1, self._final_request_number + 1))
